samcli/lib/build/app_builder.py

```diff
@@ -0,0 +1,305 @@
+
+    def build(self):
+        """
+        Build the entire application
+
+        Returns
+        -------
+        dict
+            Returns the path to where each resource was built as a map of resource's LogicalId to the path string
+        """
+
+        result = {}
+
+        for lambda_function in self._function_provider.get_all():
+
+            LOG.info("Building resource '%s'", lambda_function.name)
+            result[lambda_function.name] = self._build_function(lambda_function.name,
+                                                                lambda_function.codeuri,
+                                                                lambda_function.runtime)
+
+        return result
+
+    def update_template(self, template_dict, target_template_path, built_artifacts):
+        """
+        Given the path to built artifacts, update the template to point appropriate resource CodeUris to the artifacts
+        folder
+
+        Parameters
+        ----------
+        template_dict
+        target_template_path : str
+            Path where the template file will be written to
+
+        built_artifacts : dict
+            Map of LogicalId of a resource to the path where the the built artifacts for this resource lives
+
+        Returns
+        -------
+        dict
+            Updated template
+        """
+
+        target_dir = os.path.dirname(target_template_path)
+
+        for logical_id, resource in template_dict.get("Resources", {}).items():
+
+            if logical_id not in built_artifacts:
+                # this resource was not built. So skip it
+                continue
+
+            # Artifacts are written relative to the output template because it makes the template portability
+            #   Ex: A CI/CD pipeline build stage could zip the output folder and pass to a
+            #   package stage running on a different machine
+            artifact_relative_path = os.path.relpath(built_artifacts[logical_id], target_dir)
+
+            resource_type = resource.get("Type")
+            properties = resource.setdefault("Properties", {})
+            if resource_type == "AWS::Serverless::Function":
+                properties["CodeUri"] = artifact_relative_path
+
+            if resource_type == "AWS::Lambda::Function":
+                properties["Code"] = artifact_relative_path
+
+        return template_dict
+
+    def _build_function(self, function_name, codeuri, runtime):
+
+        config = _get_workflow_config(runtime)
+
+        # Create the arguments to pass to the builder
+
+        # Code is always relative to the given base directory.
+        code_dir = str(pathlib.Path(self._base_dir, codeuri).resolve())
+
+        # artifacts directory will be created by the builder
+        artifacts_dir = str(pathlib.Path(self._build_dir, function_name))
+
+        with osutils.mkdir_temp() as scratch_dir:
+            manifest_path = self._manifest_path_override or os.path.join(code_dir, config.manifest_name)
```

Comment:
the manifest would be requirements.txt here? for python ie

Reply:
yup, exactly